### PR TITLE
fix: Ensure correct code path for && followed by ??

### LINKED ISSUE #13614: Code path analysis for (a && b) ?? c
**Tell us about your environment**

* **ESLint Version:** v7.7.0
* **Node Version:** v12.14.0
* **npm Version:** 6.13.4

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
  parserOptions: {
    ecmaVersion: 2020
  }
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
(a && b) ?? c;
```

The rest of the bug report template isn't quite suitable for this issue.

Code path analysis gives the following:

<details>
<summary>DOT</summary>

```dot
digraph {
node[shape=box,style="rounded,filled",fillcolor=white];
initial[label="",shape=circle,style=filled,fillcolor=black,width=0.25,height=0.25];
final[label="",shape=doublecircle,style=filled,fillcolor=black,width=0.25,height=0.25];
s1_1[label="Program:enter\nExpressionStatement:enter\nLogicalExpression:enter\nLogicalExpression:enter\nIdentifier (a)"];
s1_2[label="Identifier (b)\nLogicalExpression:exit"];
s1_3[label="Identifier (c)"];
s1_4[label="LogicalExpression:exit\nExpressionStatement:exit\nProgram:exit"];
initial->s1_1->s1_2->s1_3->s1_4;
s1_1->s1_4;
s1_2->s1_4->final;
}
```
</details>

<details>
<summary>Image</summary>

image

</details>

This would be correct for `(a || b) ?? c`, but I think it isn't entirely correct for `(a && b) ?? c` since there should be a path from `a` to `c` that doesn't go through `b`:

```js
(null && console.log("b")) ?? console.log("c"); // logs only "c"
```

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[x] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

I changed the generated code path for `&&` expressions that are the left-hand operand of `??` expressions so that they show the short-circuiting of nullish values. I updated old tests that were producing incorrect results, as well.

I also did some more refactoring and added more comments with my understanding of what is happening.

Fixes #13614 

<details>
<summary>logical--and-qq</summary>

**Before:**

[GraphViz](https://dreampuf.github.io/GraphvizOnline/#digraph%20%7B%0A%20%20%20%20node%5Bshape%3Dbox%2Cstyle%3D%22rounded%2Cfilled%22%2Cfillcolor%3Dwhite%5D%3B%0A%20%20%20%20initial%5Blabel%3D%22%22%2Cshape%3Dcircle%2Cstyle%3Dfilled%2Cfillcolor%3Dblack%2Cwidth%3D0.25%2Cheight%3D0.25%5D%3B%0A%20%20%20%20final%5Blabel%3D%22%22%2Cshape%3Ddoublecircle%2Cstyle%3Dfilled%2Cfillcolor%3Dblack%2Cwidth%3D0.25%2Cheight%3D0.25%5D%3B%0A%20%20%20%20s1_1%5Blabel%3D%22Program%3Aenter%5CnExpressionStatement%3Aenter%5CnLogicalExpression%3Aenter%5CnLogicalExpression%3Aenter%5CnIdentifier%20(a)%22%5D%3B%0A%20%20%20%20s1_2%5Blabel%3D%22Identifier%20(b)%5CnLogicalExpression%3Aexit%22%5D%3B%0A%20%20%20%20s1_3%5Blabel%3D%22Identifier%20(c)%22%5D%3B%0A%20%20%20%20s1_4%5Blabel%3D%22LogicalExpression%3Aexit%5CnExpressionStatement%3Aexit%5CnProgram%3Aexit%22%5D%3B%0A%20%20%20%20initial-%3Es1_1-%3Es1_2-%3Es1_3-%3Es1_4%3B%0A%20%20%20%20s1_1-%3Es1_4%3B%0A%20%20%20%20s1_2-%3Es1_4-%3Efinal%3B%0A%7D)

graphviz(5)

**After:**

[GraphViz](https://dreampuf.github.io/GraphvizOnline/#digraph%20%7B%0A%20%20%20%20node%5Bshape%3Dbox%2Cstyle%3D%22rounded%2Cfilled%22%2Cfillcolor%3Dwhite%5D%3B%0A%20%20%20%20initial%5Blabel%3D%22%22%2Cshape%3Dcircle%2Cstyle%3Dfilled%2Cfillcolor%3Dblack%2Cwidth%3D0.25%2Cheight%3D0.25%5D%3B%0A%20%20%20%20final%5Blabel%3D%22%22%2Cshape%3Ddoublecircle%2Cstyle%3Dfilled%2Cfillcolor%3Dblack%2Cwidth%3D0.25%2Cheight%3D0.25%5D%3B%0A%20%20%20%20s1_1%5Blabel%3D%22Program%3Aenter%5CnExpressionStatement%3Aenter%5CnLogicalExpression%3Aenter%5CnLogicalExpression%3Aenter%5CnIdentifier%20(a)%22%5D%3B%0A%20%20%20%20s1_2%5Blabel%3D%22Identifier%20(b)%5CnLogicalExpression%3Aexit%22%5D%3B%0A%20%20%20%20s1_3%5Blabel%3D%22Identifier%20(c)%22%5D%3B%0A%20%20%20%20s1_4%5Blabel%3D%22LogicalExpression%3Aexit%5CnExpressionStatement%3Aexit%5CnProgram%3Aexit%22%5D%3B%0A%20%20%20%20initial-%3Es1_1-%3Es1_2-%3Es1_3-%3Es1_4%3B%0A%20%20%20%20s1_1-%3Es1_3%3B%0A%20%20%20%20s1_1-%3Es1_4%3B%0A%20%20%20%20s1_2-%3Es1_4-%3Efinal%3B%0A%7D)

graphviz(6)
</details>

<details>
<summary>logical--if-mix-and-qq-1</summary>

**Before:**

[GraphViz](https://dreampuf.github.io/GraphvizOnline/#digraph%20%7B%0A%20%20%20%20node%5Bshape%3Dbox%2Cstyle%3D%22rounded%2Cfilled%22%2Cfillcolor%3Dwhite%5D%3B%0A%20%20%20%20initial%5Blabel%3D%22%22%2Cshape%3Dcircle%2Cstyle%3Dfilled%2Cfillcolor%3Dblack%2Cwidth%3D0.25%2Cheight%3D0.25%5D%3B%0A%20%20%20%20final%5Blabel%3D%22%22%2Cshape%3Ddoublecircle%2Cstyle%3Dfilled%2Cfillcolor%3Dblack%2Cwidth%3D0.25%2Cheight%3D0.25%5D%3B%0A%20%20%20%20s1_1%5Blabel%3D%22Program%5CnIfStatement%5CnLogicalExpression%5CnLogicalExpression%5CnIdentifier%20(a)%5CnIdentifier%3Aexit%20(a)%22%5D%3B%0A%20%20%20%20s1_2%5Blabel%3D%22Identifier%20(b)%5CnIdentifier%3Aexit%20(b)%5CnLogicalExpression%3Aexit%22%5D%3B%0A%20%20%20%20s1_3%5Blabel%3D%22Identifier%20(c)%5CnIdentifier%3Aexit%20(c)%5CnLogicalExpression%3Aexit%22%5D%3B%0A%20%20%20%20s1_4%5Blabel%3D%22BlockStatement%5CnExpressionStatement%5CnCallExpression%5CnIdentifier%20(foo)%5CnIdentifier%3Aexit%20(foo)%5CnCallExpression%3Aexit%5CnExpressionStatement%3Aexit%5CnBlockStatement%3Aexit%22%5D%3B%0A%20%20%20%20s1_6%5Blabel%3D%22IfStatement%3Aexit%5CnProgram%3Aexit%22%5D%3B%0A%20%20%20%20s1_5%5Blabel%3D%22BlockStatement%5CnExpressionStatement%5CnCallExpression%5CnIdentifier%20(bar)%5CnIdentifier%3Aexit%20(bar)%5CnCallExpression%3Aexit%5CnExpressionStatement%3Aexit%5CnBlockStatement%3Aexit%22%5D%3B%0A%20%20%20%20initial-%3Es1_1-%3Es1_2-%3Es1_3-%3Es1_4-%3Es1_6%3B%0A%20%20%20%20s1_1-%3Es1_5-%3Es1_6%3B%0A%20%20%20%20s1_2-%3Es1_4%3B%0A%20%20%20%20s1_3-%3Es1_5%3B%0A%20%20%20%20s1_2-%3Es1_5%3B%0A%20%20%20%20s1_6-%3Efinal%3B%0A%7D)

graphviz(7)

**After:**

[GraphViz](https://dreampuf.github.io/GraphvizOnline/#digraph%20%7B%0A%20%20%20%20node%5Bshape%3Dbox%2Cstyle%3D%22rounded%2Cfilled%22%2Cfillcolor%3Dwhite%5D%3B%0A%20%20%20%20initial%5Blabel%3D%22%22%2Cshape%3Dcircle%2Cstyle%3Dfilled%2Cfillcolor%3Dblack%2Cwidth%3D0.25%2Cheight%3D0.25%5D%3B%0A%20%20%20%20final%5Blabel%3D%22%22%2Cshape%3Ddoublecircle%2Cstyle%3Dfilled%2Cfillcolor%3Dblack%2Cwidth%3D0.25%2Cheight%3D0.25%5D%3B%0A%20%20%20%20s1_1%5Blabel%3D%22Program%5CnIfStatement%5CnLogicalExpression%5CnLogicalExpression%5CnIdentifier%20(a)%5CnIdentifier%3Aexit%20(a)%22%5D%3B%0A%20%20%20%20s1_2%5Blabel%3D%22Identifier%20(b)%5CnIdentifier%3Aexit%20(b)%5CnLogicalExpression%3Aexit%22%5D%3B%0A%20%20%20%20s1_3%5Blabel%3D%22Identifier%20(c)%5CnIdentifier%3Aexit%20(c)%5CnLogicalExpression%3Aexit%22%5D%3B%0A%20%20%20%20s1_4%5Blabel%3D%22BlockStatement%5CnExpressionStatement%5CnCallExpression%5CnIdentifier%20(foo)%5CnIdentifier%3Aexit%20(foo)%5CnCallExpression%3Aexit%5CnExpressionStatement%3Aexit%5CnBlockStatement%3Aexit%22%5D%3B%0A%20%20%20%20s1_6%5Blabel%3D%22IfStatement%3Aexit%5CnProgram%3Aexit%22%5D%3B%0A%20%20%20%20s1_5%5Blabel%3D%22BlockStatement%5CnExpressionStatement%5CnCallExpression%5CnIdentifier%20(bar)%5CnIdentifier%3Aexit%20(bar)%5CnCallExpression%3Aexit%5CnExpressionStatement%3Aexit%5CnBlockStatement%3Aexit%22%5D%3B%0A%20%20%20%20initial-%3Es1_1-%3Es1_2-%3Es1_3-%3Es1_4-%3Es1_6%3B%0A%20%20%20%20s1_1-%3Es1_3%3B%0A%20%20%20%20s1_3-%3Es1_5-%3Es1_6%3B%0A%20%20%20%20s1_2-%3Es1_4%3B%0A%20%20%20%20s1_1-%3Es1_5%3B%0A%20%20%20%20s1_2-%3Es1_5%3B%0A%20%20%20%20s1_6-%3Efinal%3B%0A%7D)

graphviz(8)

</details>

<details>
<summary>assignment--nested-and-qq</summary>

**Before:**

[GraphViz](https://dreampuf.github.io/GraphvizOnline/#digraph%20%7B%0A%20%20%20%20node%5Bshape%3Dbox%2Cstyle%3D%22rounded%2Cfilled%22%2Cfillcolor%3Dwhite%5D%3B%0A%20%20%20%20initial%5Blabel%3D%22%22%2Cshape%3Dcircle%2Cstyle%3Dfilled%2Cfillcolor%3Dblack%2Cwidth%3D0.25%2Cheight%3D0.25%5D%3B%0A%20%20%20%20final%5Blabel%3D%22%22%2Cshape%3Ddoublecircle%2Cstyle%3Dfilled%2Cfillcolor%3Dblack%2Cwidth%3D0.25%2Cheight%3D0.25%5D%3B%0A%20%20%20%20s1_1%5Blabel%3D%22Program%3Aenter%5CnExpressionStatement%3Aenter%5CnLogicalExpression%3Aenter%5CnAssignmentExpression%3Aenter%5CnIdentifier%20(a)%22%5D%3B%0A%20%20%20%20s1_2%5Blabel%3D%22Identifier%20(b)%5CnAssignmentExpression%3Aexit%22%5D%3B%0A%20%20%20%20s1_3%5Blabel%3D%22Identifier%20(c)%22%5D%3B%0A%20%20%20%20s1_4%5Blabel%3D%22LogicalExpression%3Aexit%5CnExpressionStatement%3Aexit%5CnProgram%3Aexit%22%5D%3B%0A%20%20%20%20initial-%3Es1_1-%3Es1_2-%3Es1_3-%3Es1_4%3B%0A%20%20%20%20s1_1-%3Es1_4%3B%0A%20%20%20%20s1_2-%3Es1_4-%3Efinal%3B%0A%7D)

graphviz(10)

**After:**

[GraphViz](https://dreampuf.github.io/GraphvizOnline/#digraph%20%7B%0A%20%20%20%20node%5Bshape%3Dbox%2Cstyle%3D%22rounded%2Cfilled%22%2Cfillcolor%3Dwhite%5D%3B%0A%20%20%20%20initial%5Blabel%3D%22%22%2Cshape%3Dcircle%2Cstyle%3Dfilled%2Cfillcolor%3Dblack%2Cwidth%3D0.25%2Cheight%3D0.25%5D%3B%0A%20%20%20%20final%5Blabel%3D%22%22%2Cshape%3Ddoublecircle%2Cstyle%3Dfilled%2Cfillcolor%3Dblack%2Cwidth%3D0.25%2Cheight%3D0.25%5D%3B%0A%20%20%20%20s1_1%5Blabel%3D%22Program%3Aenter%5CnExpressionStatement%3Aenter%5CnLogicalExpression%3Aenter%5CnAssignmentExpression%3Aenter%5CnIdentifier%20(a)%22%5D%3B%0A%20%20%20%20s1_2%5Blabel%3D%22Identifier%20(b)%5CnAssignmentExpression%3Aexit%22%5D%3B%0A%20%20%20%20s1_3%5Blabel%3D%22Identifier%20(c)%22%5D%3B%0A%20%20%20%20s1_4%5Blabel%3D%22LogicalExpression%3Aexit%5CnExpressionStatement%3Aexit%5CnProgram%3Aexit%22%5D%3B%0A%20%20%20%20initial-%3Es1_1-%3Es1_2-%3Es1_3-%3Es1_4%3B%0A%20%20%20%20s1_1-%3Es1_3%3B%0A%20%20%20%20s1_2-%3Es1_4%3B%0A%20%20%20%20s1_1-%3Es1_4-%3Efinal%3B%0A%7D)

graphviz(9)

</details>

#### Is there anything you'd like reviewers to focus on?

Please double-check the digraphs on the updated tests to make sure they make sense.

### CODE REVIEW

File: lib/linter/code-path-analysis/code-path-state.js
Comment (mdjermanovic) [inline lib/linter/code-path-analysis/code-path-state.js:1]: `qqForkContext` should be renamed to `nullishForkContext` in comments on lines 124, 180, and 1219.

**Review by mdjermanovic — COMMENTED** [review]
Few minor suggestions for now.

**Review by mdjermanovic — APPROVED** [review]
Works well and all digraphs on the updated tests and a few other code examples I tried make sense. Thanks! LGTM when the minor suggestions I left are resolved.

File: lib/linter/code-path-analysis/code-path-state.js
Comment (nzakas) [reply lib/linter/code-path-analysis/code-path-state.js:1220]: Oops, right you are.

**Comment by nzakas** [comment]
Thanks, went through and fixed all that stuff.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks! Leaving this open in case anyone else would like to review it before merging.

**Review by fasttime — APPROVED** [review]
LGTM, thanks! I'll merge this in a minute.

**Pull request merged**
